### PR TITLE
Fix #2041 : Added basic slack functionality

### LINKED ISSUE #2041: Add feature for slack notification
## Current Behaviour

Currently, when a challenge is created, submission is made or a contact message is received, EvalAI Admin doesn't get any notification.

## Expected Behaviour

A slack notification should be added for all the three events.

## Deliverables
- [ ] Create a common function in `base/utils` which will take input a slack webhook URL and a message to be sent to the slack.
- [ ] Add the notification for following events:
	- [ ] When a challenge zip is uploaded
	- [ ] When a contact message is received
	- [ ] When a submission is made
- [ ] Write test cases.

## Starter Code
```
def send_slack_notification(webhook=settings.SLACK_WEBHOOKS['default'], message=""):
    '''Send slack notification to any workspace
    
    Keyword Arguments:
        webhook {string} -- slack webhook URL (default: {settings.SLACK_WEBHOOKS['default']})
        message {str} -- JSON/Text message to be sent to slack (default: {""})
    '''

    try:
        response = requests.post(
            webhook,
            data=json.dumps({'text': str(message)}),
            headers={'Content-Type': 'application/json'}
        )
    except Exception as e:
        logger.info('Exception raised while sending slack notification. \n Exception message: {}'.format(e))
```

Please feel free to ask if you have any doubts!

### PR DESCRIPTION
**Pull request by yashdusing**
Fix #2041 : First version of slack functionality is added 

Currently, working on writing the test case

### CODE REVIEW

**Comment by yashdusing** [comment]
@RishabhJain2018 Could you please verify if this is what was needed. 
Also, I haven't included any webhook in settings so by default should it be blank ?

File: apps/challenges/views.py [hunk 1]
```diff
@@ -927,15 +927,18 @@ def create_challenge_using_zip_file(request, challenge_host_team_pk):
             for email in emails:
                 user = User.objects.get(email=email)
                 host = Participant(
-                            user=user,
-                            status=Participant.ACCEPTED,
-                            team=participant_host_team,
-                    )
+                    user=user,
```
Comment (vkartik97) [inline apps/challenges/views.py]: @yashdusing 
I think the editor or Automatic PEP8 Formatter such as `autopep8` has made a lot of unnecessary changes which are not needed in this PR.

Comment (yashdusing) [reply apps/challenges/views.py]: Yeah, I’ll clean that up in a while 😅

**Comment by yashdusing** [comment]
@vkartik97 I didn’t understand what you meant by ‘without blocking the endpoint return’ . Doesn’t the ‘request’ module send HTTP request asynchronously ?

**Comment by vkartik97** [comment]
@yashdusing Please refer http://docs.python-requests.org/en/latest/user/advanced/#blocking-or-non-blocking

**Comment by vkartik97** [comment]
@yashdusing 
>With the default Transport Adapter in place, Requests does not provide any kind of non-blocking IO.

Ref: http://docs.python-requests.org/en/latest/user/advanced/#blocking-or-non-blocking

**Comment by yashdusing** [comment]
@vkartik97 Doesn’t the same thing apply for all calls to backend api as well ?

**Comment by vkartik97** [comment]
@yashdusing 
No as the endpoint has an `External Dependency` i.e Slack API and not Project Dependency. If at any point Slack API takes longer to respond, the endpoint will block the call and will proceed if there is any Success or Timeout, which would increase the overall Response Time.

**Comment by vkartik97** [comment]
@yashdusing 
No as the endpoint has an `External Dependency` i.e Slack API and not Project Level i.e Internal Dependency. If at any point Slack API takes longer to respond, the endpoint will block the call and will proceed if there is any Response or Timeout, which would increase the overall Response Time.

**Comment by yashdusing** [comment]
@vkartik97 Oh okay. Makes sense.
So, there are other libraries like requests-threads which is listed in the link you mentioned which could be used

**Comment by vkartik97** [comment]
I would suggest to first experiment with `threading.Thread`.

**Comment by yashdusing** [comment]
@deshraj @RishabhJain2018 @vkartik97 Could you please check ?

File: apps/base/utils.py [hunk 2]
```diff
@@ -69,3 +75,30 @@ def decode_data(data):
...
+            webhook,
+            data=json.dumps({'text': str(message)}),
+            headers={'Content-Type': 'application/json'}
+        )
+    except Exception as e:
+        logger.info(
+            'Exception raised while sending slack notification. \n Exception message: {}'.format(e))
+
+
+def send_slack_notification(webhook=settings.SLACK_WEBHOOKS['default'], message=""):
```
Comment (vkartik97) [inline apps/base/utils.py]: @yashdusing 
`send_slack_notification` should only run if the Project is not running in `Dev`. Add a condition for it

File: settings/common.py [hunk 3]
```diff
@@ -282,6 +282,9 @@
 # For inviting users to participant and host teams.
 ADMIN_EMAIL = "admin@cloudcv.org"
 CLOUDCV_TEAM_EMAIL = "EvalAI Team <team@cloudcv.org>"
+SLACK_WEBHOOKS = {
+    'default': ''
```
Comment (vkartik97) [inline settings/common.py]: @yashdusing 
How are you testing the feature locally if the webook is blank?

**Comment by yashdusing** [comment]
@vkartik97 I'm using a personal webhook which I haven't displayed in the PR

**Comment by yashdusing** [comment]
@vkartik97 I'll add in the condition for DEV environment in a while. I'm also trying to look for a way to return values from a thread without using global variables (for the test case) but the code works even without it.

**Comment by vkartik97** [comment]
@yashdusing 
You can use Pass by Reference to Inject a List(Mutable),  to `send_slack_notification`. Then, store the Response Code at 0th Index of List. You can then simply check if the Response Code Matches with Required Code or not for Testing, after a delay of some seconds(Preferable 2).
For Mocking POST, try `response` module.

For Example, 
```python
def threader(i):
    i[0] = 10
    print("Executed!")

i = [20]
def handle_sub_view(i):  
    threading.Thread(target=threader, args=[i]).start()
handle_sub_view(i)
time.sleep(2)
print(i[0])

# Output:
# Executed!
# 10
```

@deshraj  @RishabhJain2018  Can you please check if this method is efficient?

**Comment by deshraj** [comment]
@yashdusing @vkartik97 Thanks for all the suggestions. Karthik did point out some valid questions in the thread (thanks @vkartik97 for thinking about the corner cases). But I would recommend to first use the simple code for sending the notifications to slack without using threading. The ultimate plan is to use Celery to queue these kind of jobs and let them run in background. 

@yashdusing so the set of todos for you in this PR should be:

- [ ] Integrate slack request code shared by @RishabhJain2018
- [ ] Send slack message for different events (like: submission done, challenge zip uploaded, new user signup) 
- [ ] Mock requests for dev. so that you can write tests for this

Once we have these functionalities, the next pull request will contain the following changes:

- [ ] Integrate celery (I believe Rishabh already have a PR for that. Just needs some refining)
- [ ] Put these slack notifications in celery tasks

Please let me know if you have questions. :)

**Comment by yashdusing** [comment]
@deshraj @RishabhJain2018 Could you please review the PR ?

**Comment by yashdusing** [comment]
The test case I've written is failing because there is no webhook provided in settings. It would work correctly if a webhook is provided

**Comment by yashdusing** [comment]
@vkartik97 Correct me if I am wrong but isn't the library you provided used for mimicking the responses ?
If so, then I don't get the point of using a test case. The test case I've written is helpful only to check whether the slack API is up and the internet connection is working (since it is a simple requests usage).

**Comment by vkartik97** [comment]
@yashdusing 
- The reason of using Test is just to ensure that if in future, anyone tries to modify `send_slack_notification`, the changes should not break the feature. 
- Mocking because Unit Test is efficient when run on `Isolated Environment`, where a Fixed State(sometimes reusable) will be used for Assert Tests.

- >The test case I've written is helpful only to check whether the slack API is up and the internet connection is working

    If you are sending `Testing Slack notification functionality`  via Post Request on Webhook URL defined in Settings.py (Ref: [tests/unit/base/test_utils.py#L121](https://github.com/Cloud-CV/EvalAI/blob/5bef35b2dab06aa739c9e5b6aba2c5508f2f3122/tests/unit/base/test_utils.py#L121) ), this will be actually be used for Posting Message using `Slack API` whenever a Test is run even on `Travis`.

**Comment by yashdusing** [comment]
@deshraj @RishabhJain2018 @vkartik97 Please review the PR. I added in the mock responses for the unit tests

File: apps/base/utils.py [hunk 4]
```diff
@@ -69,3 +74,23 @@ def decode_data(data):
...
+    '''Send slack notification to any workspace
+
+    Keyword Arguments:
+        webhook {string} -- slack webhook URL (default: {settings.SLACK_WEBHOOKS['default']})
+        message {str} -- JSON/Text message to be sent to slack (default: {""})
+    '''
+    if settings.DEBUG is False:
```
Comment (deshraj) [inline apps/base/utils.py]: You don't need this check here. Please remove it. In dev also, we might want to test the slack notifications sometimes and then this condition will prohibit us from testing that.

File: apps/base/utils.py [hunk 5]
```diff
@@ -69,3 +74,23 @@ def decode_data(data):
...
+        try:
+            response = requests.post(
+                webhook,
+                data=json.dumps({'text': str(message)}),
+                headers={'Content-Type': 'application/json'}
+            )
+            return response
+        except Exception as e:
```
Comment (deshraj) [inline apps/base/utils.py]: Instead of doing exception handling, it would make more sense to check based on the response status.

File: apps/base/utils.py [hunk 6]
```diff
@@ -69,3 +74,23 @@ def decode_data(data):
...
+            logger.info(
+                'Exception raised while sending slack notification. \n Exception message: {}'.format(e))
```
Comment (deshraj) [inline apps/base/utils.py]: Please send the slack message in logger.info if the request didn't go through properly.

File: apps/challenges/views.py [hunk 7]
```diff
@@ -936,6 +936,9 @@ def create_challenge_using_zip_file(request, challenge_host_team_pk):
 
             zip_config.challenge = challenge
             zip_config.save()
+
+            send_slack_notification(
+                message="A *new challenge* has been uploaded to EvalAI")
```
Comment (deshraj) [inline apps/challenges/views.py]: This message is not informative. Please stringify a dict that contains 'link of the zip uploaded', 'primary key of uploaded config' and 'Challenge name' (if a challenge is created).

File: apps/jobs/views.py [hunk 8]
```diff
@@ -174,6 +174,7 @@ def challenge_submission(request, challenge_id, challenge_phase_id):
             submission = serializer.instance
             # publish message in the queue
             publish_submission_message(challenge_id, challenge_phase_id, submission.id)
+            send_slack_notification(message="A *new submission* has been uploaded")
```
Comment (deshraj) [inline apps/jobs/views.py]: Similarly, add the submission details here. The dict should look like this:
```
{'challenge_phase': 'Test Phase 1', 'challenge': 'Random Challenge', 'challenge_id': 2, 'event': 'New submission', 'participant': 'deshraj'}
```

File: apps/web/views.py [hunk 9]
```diff
@@ -98,6 +99,7 @@ def contact_us(request):
         if serializer.is_valid():
             serializer.save()
             response_data = {'message': 'We have received your request and will contact you shortly.'}
+            send_slack_notification(message="A *new contact message* is received")
```
Comment (deshraj) [inline apps/web/views.py]: Add the contact message details inside the slack message. Otherwise, it's not useful at all.

File: requirements/common.txt [hunk 10]
```diff
@@ -18,4 +18,5 @@ Pillow==3.4.2
 psycopg2==2.7.3.2
 PyYaml==4.2b1
 proc==0.10.1
+responses==0.5.1
```
Comment (deshraj) [inline requirements/common.txt]: Nice. This is what we were looking for. But do you think we need this in prod config? If not, then please put it in dev.txt

Comment (yashdusing) [reply requirements/common.txt]: @deshraj I'm not really sure if we need this in prod. If 'travis' uses the prod build for testing then we will need it for sure. If in the prod build, we were to remove the responses codes and put in the actual webhook then it might work for dev.txt only

**Comment by yashdusing** [comment]
@deshraj @RishabhJain2018 Could you review the PR again please ?

File: requirements/common.txt (hunk 10 above)
Comment (deshraj) [reply requirements/common.txt]: We actually don't need this in prod. Feel free to move it to dev.py

File: apps/base/utils.py (hunk 5 above)
Comment (deshraj) [reply apps/base/utils.py]: I think it should just work since Travis is also just a machine running all the tests. It should ideally work on all the machines. Can you please spend some more time on this to figure it out? Thanks for your patience.

File: settings/common.py [hunk 11]
```diff
@@ -282,6 +282,9 @@
 # For inviting users to participant and host teams.
 ADMIN_EMAIL = "admin@cloudcv.org"
 CLOUDCV_TEAM_EMAIL = "EvalAI Team <team@cloudcv.org>"
+SLACK_WEBHOOKS = {
+    'default': 'http://testslackwebhook.com/webhook'
```
Comment (deshraj) [inline settings/common.py]: Let's not hardcode it. Can you please fetch this from the environment ?

Comment (yashdusing) [reply settings/common.py]: @deshraj I didn't understand what you meant by fetching from the environment. Could you please elaborate

File: apps/base/utils.py (hunk 5 above)
Comment (yashdusing) [reply apps/base/utils.py]: Fixed this. I missed adding in mocks for test cases where send_slack_notifications was indirectly used

File: settings/common.py (hunk 11 above)
Comment (yashdusing) [reply settings/common.py]: did you mean this (https://stackoverflow.com/questions/4906977/how-do-i-access-environment-variables-from-python) ?

**Comment by yashdusing** [comment]
@deshraj @RishabhJain2018 could you please check the PR and confirm the doubts that I have ?

**Comment by yashdusing** [comment]
@deshraj @RishabhJain2018 I changed the settings so that instead of hard-coding the url, I fetched it from the OS environment

File: settings/common.py [hunk 12]
```diff
@@ -284,7 +284,7 @@
 CLOUDCV_TEAM_EMAIL = "EvalAI Team <team@cloudcv.org>"
 
 slack_webhook = os.environ.get('SLACK_WEBHOOK')
-if slack_webhook == None:
+if slack_webhook is None:
```
Comment (vkartik97) [inline settings/common.py]: @yashdusing 
Instead of this check, you can use
```python
slack_webhook = os.getenv('SLACK_WEBHOOK', 'http://testslackwebhook.com/webhook')
```
Please Ref: https://docs.python.org/3/library/os.html#os.getenv

Comment (yashdusing) [reply settings/common.py]: Thanks. Simplifies the logic :) :+1:

**Comment by RishabhJain2018** [comment]
@yashdusing Can you please resolve the conflicts? Let's get this merged 🎉

**Comment by yashdusing** [comment]
@RishabhJain2018 I've resolved the merge conflicts

**Comment by yashdusing** [comment]
@RishabhJain2018 please review

**Comment by RishabhJain2018** [comment]
@yashdusing if possible could you please resolve the conflicts here. We'd like to get it merged.
